### PR TITLE
Percent/URL Decoding Support

### PR DESCRIPTION
**Pull request by bplaxco**
### Description:

Add [percent decoding](https://en.wikipedia.org/wiki/Percent-encoding) (aka URL decoding) support.

Goals:

- Be fast
- Try not to add a lot of extra passes on the code
- Make it easier to add more formats in the future
- Nice to have: handle cases where other encodings might be partially encoded in another one[1]

[1]: For example:

This `aGVsbG8sIHdvcmxkIQ%3D%3D%0A` should decode first to `aGVsbG8sIHdvcmxkIQ==` and then `hello, world!`. But this `'hello%2C%20world%21  aGVsbG8sIHdvcmxkIQ==` shouldn't need two passes to to be fully decoded.

### TODO

* [x] Get working base implementation
* [X] Update docs
* [x] Refactor
* [x] (Extra) add hex encoding
* [x] Add tests in `detect_test.go` (make sure to test multiple levels and that the parent child relationship works properly with segments)
* [x] Add tests to confirm detection for when to decode things in multiple steps works as expected
* [x] Try to think of some edge cases to add to the tests (@zricethezav && @rgmz suggestions welcome ^_^)
* [x] Performance test, tune & refactor
* [x] Fix scan bug when scanning gitleaks repo

### Checklist:

* [x] Does your PR pass tests?
* [X] Have you written new tests for your changes?
* [x] Have you lint your code locally prior to submission?

### CODE REVIEW

File: detect/codec/encodings.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+var (
+	// encodingsRe is a regex built by combining all the encoding patterns
+	// into named capture groups so that a single pass can detect multiple
+	// encodings
+	encodingsRe *regexp.Regexp
```
Comment (zricethezav) [inline detect/codec/encodings.go:13]: this is really nifty 👍🏻

File: detect/codec/ascii.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		if !printableASCII[c] {
+			return false
+		}
+	}
+
+	return true
+}
+
+// hasByte can be used to check if a string has at least one of the provided
+// bytes. Note: make sure bs is long enough to handle the largest byte in
```
[the remark this answers is not in the document]
Comment (bplaxco) [reply detect/codec/ascii.go]: Updated ^_^

File: detect/codec/encodings.go (hunk 1 above)
Comment (bplaxco) [reply detect/codec/encodings.go:13]: Thanks! ^_^

File: detect/detect_test.go [hunk 3]
```diff
@@ -44,12 +45,43 @@ eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiY29uZmlnIjoiVzJ
...
+## one of the lines above that went through... a lot
+## and there's surrounding text around it
+Look at this value: %4EjMzMjU2NkE2MzZENTYzMDUwNTY3MDQ4%4eTY2RDcwNjk0RDY5NTUzMTRENkQ3ODYx%25%34%65TE3QTQ2MzY1NzZDNjQ0RjY1NTY3MDU5NTU1ODUyNkI2MjUzNTUzMDRFNkU0RTZCNTYzMTU1MzkwQQ== # isn't it crazy?
+## Multi percent encode two random characters close to the bounds of the base64
+## encoded data to make sure that the bounds are still correctly calculated
+secret=ZG%25%32%35%25%33%32%25%33%35%25%32%35%25%33%33%25%33%35%25%32%35%25%33%33%25%33%36%25%32%35%25%33%32%25%33%35%25%32%35%25%33%33%25%33%36%25%32%35%25%33%36%25%33%31%25%32%35%25%33%32%25%33%35%25%32%35%25%33%33%25%33%36%25%32%35%25%33%33%25%33%322RlZC1zZWNyZXQtd%25%36%64%25%34%36%25%37%33dWU=
+## The similar to the above but also touching the edge of the base64
+secret=%25%35%61%25%34%37%25%35%36jb2RlZC1zZWNyZXQtdmFsdWU%25%32%35%25%33%33%25%36%34
+## The similar to the above but also touching and overlapping the base64
+secret%3D%25%35%61%25%34%37%25%35%36jb2RlZC1zZWNyZXQtdmFsdWU%25%32%35%25%33%33%25%36%34
```
Comment (bplaxco) [inline detect/detect_test.go:84]: @zricethezav && @rgmz I think these caught one of the last bugs I was dealing with. :P

The percent encodings are intentionally not uniform depth in various spots here so traversing the predecessors needed to be able to find predecessors that overlap the adjusted value for each step up the graph and not just in the predecessors that each predecessor :P

There's a couple of spots I can short cut things a bit and I'm already doing it in one place to avoid a lot of extra checks for normal b64 & hex decoding but yeah it's actually workin' :D

**Review by zricethezav — APPROVED** [review]
At this point I'm convinced @bplaxco is some kind of wizard or time traveling AI

File: detect/codec/decoder.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+			// Adjust encodings
+			for _, p := range segment.predecessors {
+				if segment.encoded.overlaps(p.decoded) {
+					segment.encodings |= p.encodings
+				}
+			}
+		}
+
+		segments = append(segments, segment)
+		logging.Debug().Msgf(
```
Comment (zricethezav) [inline detect/codec/decoder.go:92]: love this debug message
